openvdb/points/PointMove.h

```diff
@@ -0,0 +1,1009 @@
+
+            const auto it = mTargetToSourceMap.find(leaf.pos());
+            assert(it != mTargetToSourceMap.end());
+
+            const Index sourceOffset(it->second);
+
+            const auto& array = leaf->constAttributeArray(mAttributeIndex);
+
+            PerformTypedMoveOp op(mTargetHandles, mSourceHandles,
+                leaf.pos(), sourceOffset, *leaf, offsets, moveIndices);
+            if (!processTypedArray(array, op)) {
+                this->performMove(leaf.pos(), sourceOffset, *leaf, offsets, moveIndices);
+            }
+        }
+    }
+
+private:
+    OffsetMap& mOffsetMap;
+    AttributeHandles& mTargetHandles;
+    const IndexToIndexMap& mTargetToSourceMap;
+    AttributeHandles& mSourceHandles;
+    const Index mAttributeIndex;
+    const LocalPointIndexMap& mMoveLeafMap;
+}; // struct LocalMovePointsOp
+
+} // namespace point_move_internal
+
+
+////////////////////////////////////////
+
+
+template <typename PointDataGridT, typename DeformerT>
+inline void movePoints( PointDataGridT& points, DeformerT& deformer,
+                        const std::vector<Name>& includeGroups,
+                        const std::vector<Name>& excludeGroups,
+                        future::Advect* objectNotInUse,
+                        bool threaded)
+{
+    using PointDataTreeT = typename PointDataGridT::TreeType;
+    using LeafNode = typename PointDataTreeT::LeafNodeType;
+    using LeafManagerT = typename tree::LeafManager<PointDataTreeT>;
+    using LeafRangeT = typename LeafManagerT::LeafRange;
+
+    using namespace point_move_internal;
+
+    // this object is for future use only
+    assert(!objectNotInUse);
+
+    PointDataTreeT& tree = points.tree();
+
+    // early exit if no LeafNodes
+
+    PointDataTree::LeafCIter iter = tree.cbeginLeaf();
+
+    if (!iter)      return;
+
+    // build voxel topology taking into account any point group deletion
+
+    auto newPoints = point_mask_internal::convertPointsToScalar<PointDataGrid>(
+        points, points.transform(), deformer, includeGroups, excludeGroups, threaded);
+    auto& newTree = newPoints->tree();
+
+    // create leaf managers for both trees
+
+    LeafManagerT sourceLeafManager(tree);
+    LeafManagerT targetLeafManager(newTree);
+
+    // initialize attribute handles
+    AttributeHandles sourceHandles(sourceLeafManager.leafCount());
+    AttributeHandles targetHandles(targetLeafManager.leafCount());
+
+    // map frequency to cumulative histogram
+    auto histogramLambda = [](const LeafRangeT& range) {
+        for (auto leaf = range.begin(); leaf; ++leaf) {
+            auto* buffer = leaf->buffer().data();
+            for (int i = 1; i < leaf->buffer().size(); i++) {
+                buffer[i] = buffer[i-1] + buffer[i];
+            }
+        }
+    };
+    if (threaded) {
+        tbb::parallel_for(targetLeafManager.leafRange(), histogramLambda);
+    } else {
+        histogramLambda(targetLeafManager.leafRange());
+    }
```

Comment:
should probably use the leaf manager foreach method here to avoid branching

Reply:
fair comment, I'll improve that